cookbook/rag_examples/setup_pinecone.py

```diff
@@ -0,0 +1,43 @@
+"""Script to upload a dataset to Pinecone"""
+import os
+
+import pandas as pd
+from config import EMBEDDINGS_COLUMN, PINECONE_INDEX, PINECONE_NAMESPACE
+from dotenv import load_dotenv
+from pinecone import Pinecone, ServerlessSpec
+
+load_dotenv()
+
+
+def setup_pinecone(df: pd.DataFrame) -> None:
+    """Sets up a Pincone Index and upserts embeddings from dataframe.
+
+    Each upserted embedding has ID set as index from original dataframe to help keep
+    track of the affiliated text later on. Vectors are upserted in batches of 300
+    (determined by trial and error) so as not to overload Pinecone.
+
+    Args:
+        df: The dataframe which contains embeddings to upsert.
+    """
+    pc = Pinecone(api_key=os.getenv("PINECONE_API_KEY"))
+    existing_indexes = pc.list_indexes().names()
+    if PINECONE_INDEX not in existing_indexes:
+        pc.create_index(
+            name=PINECONE_INDEX,
+            dimension=1536,
+            metric="cosine",
+            spec=ServerlessSpec(
+                cloud="aws",
+                region="us-west-2",
+            ),
+        )
+    index = pc.Index(PINECONE_INDEX)
```

Comment:
can we not just check here if the index is empty or not? Assume if empty this hasn't been run. If it isn't empty then it has been run?

Reply:
```python
# Once we've successfully uploaded the news article dataset, we don't want to
# upload it again (to save on costs). However, in the event of a failure, we want to
# make sure that we upload again so the full dataset is there.
if num_examples = index.num_examples:
    return
...
```